### PR TITLE
convert the CockroachDB controller to the controller-runtime

### PR DESCRIPTION
**Pull request by moricho**
**Description of your changes:**

This PR replaces CockroachDB cluster controller with the new controller-runtime
framework.
Although it's not directly related to this PR, I think we had better to implement `cockroachdb client` as with other controllers related to Ceph in the future to easily handle CockroachDB components.

**Which issue is resolved by this Pull Request:**
- https://github.com/rook/rook/issues/4942

**Checklist:**

- [x] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [x] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [x] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [x] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](https://github.com/rook/rook/blob/master/CONTRIBUTING.md#comments)
- [ ] Add the flag for skipping the CI if this PR does not require a build. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for more details.

[test cockroachdb]

### CODE REVIEW

**Comment by jbw976** [comment]
@moricho this is really cool that you took a stab at updating the cockroachdb operator to use controller-runtime! We are planning on taking a pass through this and offering some comments soon.  Rook's participant in the Google Summer of Code this year @prksu is going to take a pass through here because he also has a similar task to update the NFS operator to use controller-runtime #4950 😄

**Comment by moricho** [comment]
@jbw976 Thank you for your comment! I got it.

File: pkg/operator/cockroachdb/cluster/controller.go [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+	// Start object reconciliation, updating status for this
+	cluster.Status.Phase = k8sutil.ReconcilingStatus
+	err = opcontroller.UpdateStatus(r.client, cluster)
+	if err != nil {
+		return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+	}
+
+	// CREATE/UPDATE
+	reconcileResponse, err := r.reconcileCreateCluster(cluster)
```
Comment (prksu) [inline pkg/operator/cockroachdb/cluster/controller.go:182]: `controller-runtime` have utility for doing create or update object [controllerutil.CreateOrUpdate](https://github.com/kubernetes-sigs/controller-runtime/blob/a2d55b5721ce735d0f43f955bf83df87db2ce1fe/pkg/controller/controllerutil/controllerutil.go#L192)
is it possible to implement that? although in this PR we have not implemented the update yet, but in my opinion, we will be fine if use that utility.

File: pkg/operator/cockroachdb/cluster/controller.go [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+		if err != nil {
+			return reconcile.Result{}, errors.Wrap(err, "failed to set status")
+		}
+	}
+
+	// // Set a finalizer so we can do cleanup before the object goes away
+	// err = opcontroller.AddFinalizerIfNotPresent(r.client, cluster)
+	// if err != nil {
+	// 	return reconcile.Result{}, errors.Wrap(err, "failed to add finalizer")
+	// }
```
[the remark this answers is not in the document]
Comment (prksu) [reply pkg/operator/cockroachdb/cluster/controller.go:149]: https://github.com/rook/rook/blob/ab8e6d26856338da7e52516aaea2563e2ca98aa4/pkg/operator/cockroachdb/cluster/controller.go#L160-L163

**Comment by leseb** [comment]
@moricho any plan to pursue this work? Thanks

**Comment by leseb** [comment]
Closing this due to inactivity. Feel free to re-open. Thanks.

**Pull request closed without merging**